emacs/lisp/init-vulpea.el

```diff
@@ -0,0 +1,339 @@
+;;; init-vulpea.el --- Note taking and project management  -*- lexical-binding: t; -*-
+;;
+;; Copyright (c) 2015-2021 Boris Buliga
+;;
+;; Author: Boris Buliga <boris@d12frosted.io>
+;; Maintainer: Boris Buliga <boris@d12frosted.io>
+;; Version: 0.1
+;; Package-Requires: ((emacs "27.1"))
+;;
+;; Created: 08 Feb 2021
+;;
+;; URL: https://github.com/d12frosted/environment/emacs
+;;
+;; License: GPLv3
+;;
+;; This program is free software; you can redistribute it and/or
+;; modify it under the terms of the GNU General Public License as
+;; published by the Free Software Foundation, either version 3 of the
+;; License, or (at your option) any later version.
+;;
+;; This program is distributed in the hope that it will be useful, but
+;; WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+;; General Public License for more details.
+;;
+;; You should have received a copy of the GNU General Public License
+;; along with this program. If not, see
+;; <http://www.gnu.org/licenses/>.
+;;
+;; This file is not part of GNU Emacs.
+;;
+;;; Commentary:
+;;
+;; This module configures all things for note taking and project
+;; management. The stars are `org' and `org-roam'.
+;;
+;;; Code:
+
+(require 'init-path)
+(require 'init-elpa)
+(require 'init-env)
+(require 'init-selection)
+(require 'config-vulpea)
+
```

Comment:
Seems that these libs need to be required here, or they are not called:

(require 'lib-vulpea)
(require 'lib-vulpea-agenda)
(require 'lib-vulpea-buffer-prop)
(require 'lib-vulpea-capture)
(require 'lib-vulpea-id)
(require 'lib-vulpea-refile)

Reply:
The idea is that you don't want to `require` them directly, otherwise lazy loading of `org` (and some other packages) will become eager :) Ideally, `lib-*` files should not be required directly, but be loaded via autoloads, which are generated by 

https://github.com/d12frosted/environment/blob/d6756ff5ed8f1460ab8bebd7ac9f157fcc582997/emacs/Makefile#L8

They are being loaded in `init.el`:

https://github.com/d12frosted/environment/blob/d6756ff5ed8f1460ab8bebd7ac9f157fcc582997/emacs/init.el#L63

That's the core idea of this revamp. I want to get proper linting and byte compilation **and** lazy loading.

In general, I think I should make it clear which lib file is OK to require and which should be autoloaded. Maybe something like `lib-vulpea` is safe to require, but `autoloads-vulpea` should be autoloaded. WDYT?

Reply:
Forgot to say, if you are trying to run these configs, please follow `How to run` section of this PR description. Soon I will provide normal README.

Reply:
Oops! I definitively have ignored the `How to run` section and kickstarted emacs directly, and that might be the reason for failing to load the `lib-*` files lazily:

`Error (use-package): org-agenda/:config: Symbol’s value as variable is void: vulpea-agenda-cmd-refile
Error (use-package): org-capture/:config: Symbol’s function definition is void: vulpea-capture-setup
`

My bad and kinda embarrassed :(.

I will experiment with it a little bit after setting up the `eru.sh`

Really appreciate for pointing it out.

Reply:
Yep, that would be the reason for those errors. Initially I silenced `autoloads` file loading, because of the bootstrapping flow. But after some changes, I can throw an error now if the autoloads file is not present.

> My bad and kinda embarrassed :(.

Oh don't feel embarrassed! This is the reason I dislike implicit things, it's not clear what's happening unless you **know**. 

> I will experiment with it a little bit after setting up the eru.sh

`eru` is kind of opinionated, but `eru.sh install emacs` should be fine for most users. You can also fallback to manual installation :)

Reply:
Please see aeb1794d. Now you would get an error during startup :smile_cat:

Reply:
Ah Yessss! After certain tweaks, my Emacs config refuses to work now. It may be due to changed working directories though, and that's probably on my part.

Still trying to get the omnipotent `eru.sh ` running.  Our environments vary so much and I know it is gonna cost me one thousand and one nights 🤣, but it's worth a try.

Reply:
> It may be due to changed working directories though, and that's probably on my part.

If you have the code pushed somewhere, I can take a look.

> Still trying to get the omnipotent eru.sh running. Our environments vary so much and I know it is gonna cost me one thousand and one nights 🤣, but it's worth a try.

Don't hesitate to share any issues you are running into. Maybe I can help. Would be also nice to hear what are the differences 😸 

If you wish, you can contact me via email - boris at d12frosted dot io.

Reply:
Hi @d12frosted,

Sorry for the 2-day silence, don't wanna bother you with my nonsense this early with a messy brain.

The good news is, I finally got the eru.sh and eldev working. So far so good.

> If you wish, you can contact me via email.

Thanks for offering help. And I still need to update and declutter my configs.

Reply:
Now I have to be sorry for 4 day silence 😄 

> don't wanna bother you with my nonsense this early with a messy brain.

Don't worry about that 😸 

> The good news is, I finally got the eru.sh and eldev working. So far so good.

Hooray! Happy to hear.

> I believe I will need your help sometime.

Don't hesitate to contact me.

---

Few important updates:

1. I switched back to `straight.el` as `quelpa` + `package.el` where not reliable with regards to upgrading packages. Most of the time when I wanted to upgrade something I had to nuke `path-packages-dir`. Also thanks to `straight.el` startup time is lower (e.g. Emacs loads faster x1.5 times).
2. I've added README file explaining the goals and approach.
3. I've implemented/migrated the most important stuff, so now I am ready to merge it into the master. Will give it day or two and that's it 😸